### PR TITLE
Use ruff linter and formatter

### LINKED ISSUE #162: Pre-commit for sorting imports
@jwallwork23 I saw now that `.utility` is separated from other relative imports. And imports are a bit disordered in general. 

Perhaps we should (in a separate issue) organise imports manually or e.g. add [isort](https://pycqa.github.io/isort/index.html) to pre-commit?

_Originally posted by @ddundo in https://github.com/mesh-adaptation/goalie/pull/161#discussion_r1564863797_

### PR DESCRIPTION
**Pull request by ddundo**
Closes #162.

This PR introduces ruff for linting and formatting. The relevant changes are done in `pyproject.toml`, `.pre-commit-config.yaml`, `Makefile` and `.flake8` (deleted). Other changes are due to applying the pre-commit, which sorts the imports: first built-ins, then any installed packages, and finally goalie and its modules.

I temporarily disabled "B" and "C" linting rules as they were blocking the push. See #165.

### CODE REVIEW

**Comment by ddundo** [comment]
Reordering the imports has now caused some errors to be raised. I guess because of the `from firedrake import *`. I'll investigate in the next few days.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,10 +1,6 @@
 repos:
--   repo: https://github.com/psf/black
-    rev: stable
+-   repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.3.7  # ruff version
```
Comment (jwallwork23) [inline .pre-commit-config.yaml:3]: This comment is redundant, please drop it. (See https://github.com/mesh-adaptation/mesh-adaptation-docs/wiki/Development-Practices#code-comments.)

**Review by jwallwork23 — APPROVED** [review]
All makes sense, thanks @ddundo. Yes, please remove the Black badge from the README.

Aside from that just one minor comment - feel free to merge once you've addressed these.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (ddundo) [reply .pre-commit-config.yaml]: Done in e037db64f41eee5ac7cfa541c09babcce8bf8679. Also removed the Black badge.

**Pull request merged**